### ISSUE TITLE
[BUG] `step_and_cost` not updating parameters with JAX circuit

### ISSUE THREAD

**Issue by cvsik** [PennyLaneAI/pennylane#2614]
### Expected behavior

The following minimal code snippet (see below, copied mostly from the JAX [tutorial](https://pennylane.ai/qml/demos/tutorial_jax_transformations.html)) should update the circuit parameter
using the optimizer's `step_and_cost` function. Manually computing the gradient yields a non-zero result, so the parameter should definitely change when applying `step_and_cost`.

### Actual behavior

`step_and_cost` does not update the circuit parameter:

```
before 0.1234 0.9923958767048912
after 0.1234 0.9923958767048912
gradient -0.12308705821137626
```

I assume that I'm doing something wrong in the setup of `step_and_cost`, but there is no error/warning about a possibly erroneous setup.

### Source code

```Python
from jax.config import config
config.update("jax_enable_x64", True)

import jax
import jax.numpy as jnp
import pennylane as qml

dev = qml.device("default.qubit", wires=2)


@qml.qnode(dev, interface="jax")
def circuit(param):
    qml.RX(param, wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0))


grad_circuit = jax.grad(circuit)

# initial value
param = 0.1234
print("before", param, circuit(param))

opt = qml.GradientDescentOptimizer(stepsize=0.4)
# try to make one step
param, val_prev = opt.step_and_cost(circuit, param, grad_fn=grad_circuit)
print("after", param, val_prev)
# manual gradient computation
print("gradient", grad_circuit(param))
```

### System information

```shell
Name: PennyLane
Version: 0.24.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/cvsik/Projects/covqcstack/qcware/pennylane
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, retworkx, scipy, semantic_version, toml
Required-by: covvqetools, PennyLane-Lightning

Platform info:           Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.10
Python version:          3.8.13
Numpy version:           1.22.3
Scipy version:           1.8.0
Installed devices:
- lightning.qubit (PennyLane-Lightning-0.23.0)
- default.gaussian (PennyLane-0.24.0.dev0)
- default.mixed (PennyLane-0.24.0.dev0)
- default.qubit (PennyLane-0.24.0.dev0)
- default.qubit.autograd (PennyLane-0.24.0.dev0)
- default.qubit.jax (PennyLane-0.24.0.dev0)
- default.qubit.tf (PennyLane-0.24.0.dev0)
- default.qubit.torch (PennyLane-0.24.0.dev0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by josh146** [comment]
Thanks @cvsik for catching this! I believe this is a subtlety of the PennyLane optimizers. They were written primarily to work for the Autograd interface and not JAX; they work with JAX and `grad_fn` due to the very similar APIs between JAX and Autograd.

However, a consequence of these optimizers being written directly for Autograd is that the `step_and_cost` method determines the trainable parameters to update by looking for the `requires_grad` attribute.

This allows a hotfix, though: I just tested it, and it seems that specifying `requires_grad` on the trainable parameter, e.g.,

```python
param = qml.numpy.array(0.1234, requires_grad=True)
```

allows your example to work:

```
before 0.1234 0.9923958767048912
after 0.1726348232845505 0.9923958767048912
gradient -0.1717786004986339
```
